core/engine/src/value/conversions/serde_json.rs

```diff
@@ -113,6 +113,15 @@ impl JsValue {
     ///
     /// Panics if the `JsValue` is `Undefined`.
     pub fn to_json(&self, context: &mut Context) -> JsResult<Value> {
+        let mut stack = vec![];
+        self.to_json_inner(context, &mut stack)
+    }
+
+    fn to_json_inner(
+        &self,
+        context: &mut Context,
+        stack: &mut Vec<JsObject<dyn NativeObject>>,
```

Comment:
Suggestion: You could use [`indexmap::IndexSet`](https://docs.rs/indexmap/latest/indexmap/set/struct.IndexSet.html) (already a dependency) to speed up the check if an object is with the stack. It's set that preserves the order of inserted object so you can call `.pop()` for popping the last inserted element.

Reply:
Thanks! I thought the reason of using vec in JSON.stringify would be JSObject not being hashable, but I just realised that it implements `Hash`. We can just use a pure set/map as well (not sure IndexSet or std HashSet is faster), but I'm using IndexSet here anyway as you suggested.

Reply:
yes, regular `HashSet` could be used too (and it's a better choice), for some reason I though that the order mattered because it was a vec before. :sweat_smile:

Reply:
Updated!